k2/csrc/connect.cu

```diff
@@ -0,0 +1,497 @@
+                  *fsas_row_splits1_data = fsas_.RowSplits(1).Data(),
+                  *dest_states_data = dest_states_.values.Data();
+    char *keep_arc_data = arc_renumbering.Keep().Data();
+    int32_t *next_iter_states_data = next_iter_states.Data();
+    K2_EVAL(
+        c_, arcs_shape.NumElements(), lambda_set_arc_renumbering,
+        (int32_t arcs_idx012)->void {
+          // note: the prefix `arcs_` means it is an idxXXX w.r.t. `arcs_shape`.
+          // the prefix `fsas_` means the variable is an idxXXX w.r.t. `fsas_`.
+          int32_t arcs_idx01 = arcs_row_ids2_data[arcs_idx012],
+                  arcs_idx01x = arcs_row_splits2_data[arcs_idx01],
+                  arcs_idx2 = arcs_idx012 - arcs_idx01x,
+                  fsas_idx01 = states_data[arcs_idx01],  // a state index
+                  fsas_idx01x = fsas_row_splits2_data[fsas_idx01],
+                  fsas_idx012 = fsas_idx01x + arcs_idx2,
+                  fsas_dest_state_idx01 = dest_states_data[fsas_idx012];
+          // if this arc is a self-loop, just ignore this arc as we won't
+          // processe the dest_state (current state) again
+          if (fsas_dest_state_idx01 == fsas_idx01 ||
+              accessible_data[fsas_dest_state_idx01]) {
+            keep_arc_data[arcs_idx012] = 0;
+            return;
+          }
+          keep_arc_data[arcs_idx012] = 1;
+          next_iter_states_data[arcs_idx012] = fsas_dest_state_idx01;
+        });
+
+    Array1<int32_t> new2old_map = arc_renumbering.New2Old();
+    if (new2old_map.Dim() == 0) {
+      // There are no new states.  This means we terminated.  We'll check from
+      // calling code that we processed all arcs.
+      return nullptr;
+    }
+    // `new_states` will contain state-ids which are idx01's into `fsas_`.
+    Array1<int32_t> new_states = next_iter_states[new2old_map];
+    Array1<int32_t> new_states_row_ids(c_, new_states.Dim());  // will map to
+                                                               // FSA index
+    const int32_t *new2old_map_data = new2old_map.Data();
+    int32_t *new_states_row_ids_data = new_states_row_ids.Data();
+    K2_EVAL(
+        c_, new_states.Dim(), lambda_set_row_ids,
+        (int32_t new_state_idx)->void {
+          int32_t arcs_idx012 = new2old_map_data[new_state_idx],
+                  arcs_idx01 = arcs_row_ids2_data[arcs_idx012],  // state index
+              arcs_idx0 = arcs_row_ids1_data[arcs_idx01];        // FSA index
+          new_states_row_ids_data[new_state_idx] = arcs_idx0;
+        });
+
+    int32_t num_fsas = fsas_.Dim0();
+    Array1<int32_t> new_states_row_splits(c_, num_fsas + 1);
+    RowIdsToRowSplits(new_states_row_ids, &new_states_row_splits);
+
+    auto ans = std::make_unique<Ragged<int32_t>>(
+        RaggedShape2(&new_states_row_splits, &new_states_row_ids, new_states.Dim()),
+        new_states);
+    // The following will ensure the answer has deterministic numbering
+    SortSublists(ans.get());
```

Comment:
I think there's nothing about `deterministic` here.  What we need in `GetNextBatch` iteration is just `accessible_`, so I belive we can remove this line.

Reply:
Yes, No need to sort. I copied from topsort, should remove.